[jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28985 [quant][graphmode] Remove observer module after insert_quant_dequant
* **#29048 [jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object**
* #28984 [jit] Add removeAttribute to `ClassType`

Summary:
In order to support remove_attribute in module, we need to support
remove slot in ivalue::Object, it's caller's responsiblity to gaurantee
the safety of the remove operation

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18343464](https://our.internmc.facebook.com/intern/diff/D18343464)

jerryzh168 committed:
[jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object

Summary:
In order to support remove_attribute in module, we need to support
remove slot in ivalue::Object, it's caller's responsiblity to gaurantee
the safety of the remove operation

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object"


Summary:
In order to support remove_attribute in module, we need to support
remove slot in ivalue::Object, it's caller's responsiblity to gaurantee
the safety of the remove operation

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object"


Summary:
In order to support remove_attribute in module, we need to support
remove slot in ivalue::Object, it's caller's responsiblity to gaurantee
the safety of the remove operation

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -199,6 +199,12 @@ void ivalue::Object::setAttr(const std::string& name, IValue v) {
+void ivalue::Object::unsafeRemoveAttr(const std::string& name) {
+  const size_t slot = type_.type_->getAttributeSlot(name);
+  unsafeRemoveSlot(slot);
+  type_.type_->unsafeRemoveAttribute(name);
```
This won't work in the case where there are multiple objects that have the same type. The first time, it will remove it from the type, then subsequent times will attempt to remove it and fail. This API should just remove a particular slot. `unsafeRemoveSlot`

jerryzh168 replied:
how about only allow remove attr if the type is used by one object? I don't feel it make much sense to remove the attribute of an object and don't remove the attribute in its underlying type.

jerryzh168 replied:
never mind, looks like use_count doesn't help me to find out how many objects uses this type.

zdevito reviewed (commented):
See the inline comment -- this needs to work only on the object to not hit corner cases for multiple objects.

jerryzh168 committed:
Update on "[jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object"


Summary:
In order to support remove_attribute in module, we need to support
remove slot in ivalue::Object, it's caller's responsiblity to gaurantee
the safety of the remove operation

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object"


Summary:
In order to support remove_attribute in module, we need to support
remove slot in ivalue::Object, it's caller's responsiblity to gaurantee
the safety of the remove operation

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object"


Summary:
In order to support remove_attribute in module, we need to support
remove slot in ivalue::Object, it's caller's responsiblity to gaurantee
the safety of the remove operation

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
